src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,153 @@
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(outOfBoundsMsg(index));
+        }
+    }
+
+    private String outOfBoundsMsg(int index) {
+        return "Index: " + index + ", Size: " + size;
+    }
+
+    private Object[] growIfArrayFull(int minCapacity) {
+        int oldCapacity = elementData.length;
+        if (oldCapacity > 0 || elementData != DEFAULTCAPACITY_EMPTY_ELEMENTDATA) {
+            int newCapacity = (int)(oldCapacity * 1.5);
+            Object[] elementDataNew = new Object[newCapacity];
+            System.arraycopy(elementData,0, elementDataNew, 0, elementData.length);
+            return elementDataNew;
+        } else {
+            return elementData = new Object[Math.max(DEFAULT_CAPACITY, minCapacity)];
+        }
+    }
+
+    private Object[] growIfArrayFull() {
+        return growIfArrayFull(size + 1);
+    }
+
     @Override
     public void addAll(List<T> list) {
-
+        Object[] elementDataNew = new Object[list.size()];
+        for (int i = 0; i < elementDataNew.length; i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index <= size - 1 && index >= 0) {
+            return (T) elementData[index];
+        } else {
+            throw new ArrayListIndexOutOfBoundsException(outOfBoundsMsg(index));
+        }
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index <= size - 1 && index >= 0) {
+            elementData[index] = value;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException(outOfBoundsMsg(index));
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException(outOfBoundsMsg(index));
+        }
+        final int newSize = size - 1;
+        T element = (T) elementData[index];
+        if (newSize > index) {
+            System.arraycopy(elementData, index + 1, elementData, index, newSize - index);
+        }
+        size = newSize;
+        return element;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        T removedElement = null;
+        int index = NEGATIVE_INDEX_MARKER;
+        for (int i = 0; i < size; i++) {
+            if (elementData[i] != null && elementData[i].equals(element)) {
+                index = i;
+                removedElement = (T) elementData[i];
+                break;
+            }
+
+            if (element == null && elementData[i] == null) {
+                size--;
+                return null;
+            }
+        }
+
+        if (index == NEGATIVE_INDEX_MARKER) {
+            throw new NoSuchElementException();
+        }
+        final int newSize = size - 1;
+        System.arraycopy(elementData, index + 1, elementData, index, newSize - index);
+        size = newSize;
+        return removedElement;
```

Comment:
Don't create repeating code. The logic for removing an element by index is repeated. Consider refactoring the common logic into a separate method to avoid duplication.